crates/ty_python_semantic/src/types/call.rs

```diff
@@ -1,14 +1,87 @@
 use super::context::InferContext;
-use super::{Signature, Type};
+use super::{Signature, Type, TypeContext};
 use crate::Db;
 use crate::types::PropertyInstanceType;
 use crate::types::call::bind::BindingError;
+use ruff_python_ast as ast;
 
 mod arguments;
 pub(crate) mod bind;
 pub(super) use arguments::{Argument, CallArguments};
 pub(super) use bind::{Binding, Bindings, CallableBinding, MatchedArgument};
 
+impl<'db> Type<'db> {
+    pub(crate) fn try_call_bin_op(
+        db: &'db dyn Db,
+        left_ty: Type<'db>,
+        op: ast::Operator,
+        right_ty: Type<'db>,
+    ) -> Result<Bindings<'db>, CallBinOpError> {
+        // We either want to call lhs.__op__ or rhs.__rop__. The full decision tree from
+        // the Python spec [1] is:
+        //
+        //   - If rhs is a (proper) subclass of lhs, and it provides a different
+        //     implementation of __rop__, use that.
+        //   - Otherwise, if lhs implements __op__, use that.
+        //   - Otherwise, if lhs and rhs are different types, and rhs implements __rop__,
+        //     use that.
+        //
+        // [1] https://docs.python.org/3/reference/datamodel.html#object.__radd__
```

Comment:
I extracted this from `builder.rs`. The lines are unchanged except that the `map(|outcome| outcome.return_type())` remains in `builder.rs`